mlir/lib/Dialect/Rock/Transforms/AccelEmitter.cpp

```diff
@@ -0,0 +1,431 @@
+                          {gridSize / gStride, gStride / nBlocks, nBlocks});
+  splitMemoryCoords.merge(
+      {"wave", "m_tid", "n_tid"}, {3, 4, 5}, "tid",
+      {wavesInKernelBlock, waveSize / inputSpanLen, inputSpanLen});
+  splitMemoryCoords.merge(
+      {"i", "j", "vec_group", "vec_item"}, {6, 7, 8, 9}, "item",
+      {numElements / (blocksPerRepeat * rowGroupsPerBlock * rowGroupSize),
+       blocksPerRepeat, rowGroupsPerBlock, rowGroupSize});
+  TransformMapAttr splitMemoryCoordsAttr = splitMemoryCoords.get();
+
+  // "blkMajor" and "blkMinor" are placeholder names because we don't know
+  // if they'll be column or row until we check for broadcast-ness.
+  auto toRowsAndCols =
+      TopDownTMBuilder::below(splitMemoryCoords, splitMemoryCoordsAttr);
+  llvm::StringMap<uint32_t> rowsAndColsIdxs =
+      expandNamesInPlace(splitMemoryCoords, {{"wave", {"wave_m", "wave_n"}},
+                                             {"i", {"m_i", "n_i"}},
+                                             {"j", {"blkMajor", "blkMinor"}}});
+  TopDownTMBottomDimsWrapper rowsAndColsWrap(toRowsAndCols, rowsAndColsIdxs);
+  rowsAndColsWrap.passThrough({"g", "m", "n"});
+  rowsAndColsWrap.merge({"wave_m", "wave_n"}, "wave",
+                        {wavesInKernelBlock / nWaves, nWaves});
+  rowsAndColsWrap.passThrough({"m_tid", "n_tid"});
+  rowsAndColsWrap.merge({"m_i", "n_i"}, "i",
+                        {splitMemoryCoords.endSize("i") / nRepeats, nRepeats});
+
+  // Here we use the full builder API since we want index and name control
+  bool isABroadcast = (nPerRepeat >= mPerRepeat);
+  SmallVector<StringRef, 2> rowsFirst = {"blk_row", "blk_col"};
+  SmallVector<StringRef, 2> colsFirst = {"blk_col", "blk_row"};
+  toRowsAndCols.merge(
+      isABroadcast ? rowsFirst : colsFirst,
+      {rowsAndColsIdxs["blkMajor"], rowsAndColsIdxs["blkMinor"]}, "j",
+      {splitMemoryCoords.endSize("j") / blocksInOutRegs, blocksInOutRegs});
+  toRowsAndCols.passThrough(
+      {"vec_group", "vec_item"},
+      {rowsAndColsIdxs["vec_group"], rowsAndColsIdxs["vec_item"]},
+      {"vec_group", "vec_item"});
+
+  TransformMapAttr toRowsAndColsAttr = toRowsAndCols.get();
+
+  auto toMatrixC = TopDownTMBuilder::below(toRowsAndCols, toRowsAndColsAttr);
+  toMatrixC.passThrough({"gemmG"}, {0}, {"g"});
+
+  toMatrixC.embed(
+      "gemmM", 1, M,
+      {"m", "wave_m", "m_tid", "m_i", "blk_row", "vec_group", "vec_item"},
+      {mPerBlock, mPerWave, rowGroupSize, mPerRepeat, m,
+       inputSpansPerMfmaIn * rowGroupSize, 1});
+  toMatrixC.embed("gemmN", 2, N, {"n", "wave_n", "n_i", "blk_col", "n_tid"},
+                  {nPerBlock, nPerWave, nPerRepeat, n, 1});
+  TransformMapAttr toMatrixCAttr = toMatrixC.get();
+
+  ArrayAttr idToMatrixCMaps =
+      b.getArrayAttr({splitMemoryCoordsAttr, toRowsAndColsAttr, toMatrixCAttr});
+  return idToMatrixCMaps;
+}
+
+Value MfmaEmitter::computeOutputConversion(PatternRewriter &b, Location loc,
```

Comment:
Note to future us: if we ever do the "output type conversions as fusions" bit, we may not need this.